src/main/java/hudson/remoting/Engine.java

```diff
@@ -247,6 +250,16 @@ public void run() {
                         }
                         host = con.getHeaderField("X-Jenkins-JNLP-Host"); // controlled by hudson.TcpSlaveAgentListener.hostName
                         if (host == null) host=url.getHost();
+                        String names = con.getHeaderField("X-Jenkins-Agent-Protocols");
+                        agentProtocolNames = new HashSet<String>();
+                        if (names != null) {
+                            for (String name: names.split(",")) {
+                                name = name.trim();
+                                if (!name.isEmpty()) {
```

Comment:
Because StringUtils is not on the classpath in remoting.jar (which is supposed to be self-contained)